src/NuGet.Clients/NuGet.CommandLine/MsBuildToolsetEx.cs

```diff
@@ -0,0 +1,174 @@
+using System.Linq;
+using System.Runtime.InteropServices.ComTypes;
+using Microsoft.Build.Evaluation;
+using Microsoft.VisualStudio.Setup.Configuration;
+
+namespace NuGet.CommandLine
+{
+    /// <summary>
+    /// "Extended" type to include Toolset from the classic installation story and the appropriate values from
+    /// ISetupInstance in the new (with MSBuild v15.1) installation story. Allows all discovered installations
+    /// to be contained in the same collection.
+    /// </summary>
+    public class MsBuildToolsetEx: IComparable
+    {
+        private readonly Toolset _toolset;
+        private readonly DateTime _installDateTime = DateTime.MinValue;
+
+        public MsBuildToolsetEx(Toolset toolset)
+        {
+            _toolset = toolset;
+
+            Version parsedToolsVersion;
+            if (!Version.TryParse(_toolset?.ToolsVersion, out parsedToolsVersion))
+            {
+                parsedToolsVersion = new Version(0, 0);
+            }
+            ParsedToolsVersion = parsedToolsVersion;
+
+            ToolsPath = GetMSBuildPathFromVsPath(_toolset?.ToolsPath);
+            if (string.IsNullOrEmpty(ToolsPath))
+            {
+                ToolsPath = _toolset?.ToolsPath; // This fallback will service pre-v15.1 toolsets
+            }
+        }
+
+        public MsBuildToolsetEx(Toolset toolset, DateTime installDateTime): this(toolset)
+        {
+            _installDateTime = installDateTime;
+        }
+
+        internal MsBuildToolsetEx(ISetupInstance sxsToolset): this(
+            new Toolset(toolsVersion: sxsToolset.GetInstallationVersion(),
+                        toolsPath: sxsToolset.GetInstallationPath(),
+                        projectCollection: new ProjectCollection(),
+                        msbuildOverrideTasksPath: string.Empty))
+        {
+            _installDateTime = ConvertFILETIMEToDateTime(sxsToolset.GetInstallDate());
+        }
+
+        internal MsBuildToolsetEx(string toolsVersion, string toolsPath): this(
+            new Toolset(toolsVersion: toolsVersion,
+                        toolsPath: toolsPath,
+                        projectCollection: new ProjectCollection(),
+                        msbuildOverrideTasksPath: string.Empty))
+        {
+        }
+
+        public string ToolsVersion
+        {
+            get
+            {
+                return _toolset?.ToolsVersion;
+            }
+        }
+
+        public Version ParsedToolsVersion { get; private set; }
+
+        private static Version SafeParseVersion(string version)
+        {
+            Version result;
+
+            if (Version.TryParse(version, out result))
+            {
+                return result;
+            }
+            else
+            {
+                return new Version(0, 0);
+            }
+        }
+
+        public string ToolsPath { get; private set; }
+
+        public static IEnumerable<MsBuildToolsetEx> AsMsToolsetExCollection(IEnumerable<Toolset> toolsets)
```

Comment:
`AsMsBuildToolsetExEnumerable`?  System.Core.dll uses [`AsEnumerable`](https://referencesource.microsoft.com/#System.Core/System/Linq/Enumerable.cs,3ee4ab000402fbeb) for returning an `IEnumerable<T>`.